File: eeg-backend/neuro_yogic/aperiodic.py
"""
aperiodic.py
============
Aperiodic (1/f) spectral decomposition — the reference oracle for the C# port in
``src/NeuroYogic.SignalProcessing/Dsp/Aperiodic.cs``.

Motivation
----------
A neural power spectrum is the sum of rhythmic *oscillatory* peaks (alpha, beta,
…) riding on a *non-oscillatory* 1/f background. That background is not noise:
its steepness (the aperiodic **exponent**) tracks excitation/inhibition balance
and arousal, and it shifts in deep meditative absorption in ways band-ratios
conflate away. We parameterise it FOOOF-style:

    log10(PSD) ≈ offset − exponent · log10(f)

A single robust least-squares line over 2–40 Hz gives (exponent, offset). This
is the simple single-pass estimate; a peak-masked refit can follow if needed.
Computed per channel then averaged, matching the band-power convention.
"""
import numpy as np


def aperiodic_fit(freqs, psd, lo: float = 2.0, hi: float = 40.0) -> dict:
    """Fit the 1/f background over [lo, hi] Hz.

    Parameters
    ----------
    freqs : ndarray (n_freq,)
    psd   : ndarray (n_channels, n_freq) or (n_freq,)
    lo,hi : fit band in Hz (inclusive)

    Returns
    -------
    dict with keys: exponent (= −slope of the log-log line), offset (intercept).
    """
    freqs = np.asarray(freqs, dtype=np.float64)
    psd = np.atleast_2d(np.asarray(psd, dtype=np.float64))
    mask = (freqs >= lo) & (freqs <= hi) & (freqs > 0.0)
    f = freqs[mask]
    if f.size < 2:
        return {"exponent": 0.0, "offset": 0.0}

    logf = np.log10(f)
    exps, offs = [], []
    for ch in psd:
        logp = np.log10(np.maximum(ch[mask], 1e-20))
        slope, intercept = _lin_fit(logf, logp)
        exps.append(-slope)      # aperiodic exponent = negative log-log slope
        offs.append(intercept)
    return {"exponent": float(np.mean(exps)), "offset": float(np.mean(offs))}


def _lin_fit(xs, ys):
    """Ordinary-least-squares slope & intercept (kept identical to the C# port)."""
    n = len(xs)
    mx = float(np.sum(xs)) / n
    my = float(np.sum(ys)) / n
    num = 0.0
    den = 0.0
    for i in range(n):
        dx = xs[i] - mx
        num += dx * (ys[i] - my)
        den += dx * dx
    slope = num / den if den != 0.0 else 0.0
    intercept = my - slope * mx
    return slope, intercept

File: eeg-backend/src/NeuroYogic.SignalProcessing/Dsp/Aperiodic.cs
using NeuroYogic.Domain.Analysis;

namespace NeuroYogic.SignalProcessing.Dsp;

/// <summary>
/// Aperiodic (1/f) spectral decomposition — faithful C# port of the Python
/// oracle in <c>neuro_yogic/aperiodic.py</c>. Fits the non-oscillatory 1/f
/// background of the power spectrum, FOOOF-style:
/// <c>log10(PSD) ≈ offset − exponent · log10(f)</c> over 2–40 Hz, per channel
/// then averaged. The exponent tracks excitation/inhibition balance and arousal.
/// </summary>
internal static class Aperiodic
{
    public static AperiodicFit Fit(double[] freqs, double[][] psd, double lo = 2.0, double hi = 40.0)
    {
        var maskIdx = new List<int>();
        for (var i = 0; i < freqs.Length; i++)
            if (freqs[i] >= lo && freqs[i] <= hi && freqs[i] > 0.0) maskIdx.Add(i);
        if (maskIdx.Count < 2) return new AperiodicFit(0.0, 0.0);

        var logf = new double[maskIdx.Count];
        for (var j = 0; j < maskIdx.Count; j++) logf[j] = Math.Log10(freqs[maskIdx[j]]);

        var nCh = psd.Length;
        double expSum = 0.0, offSum = 0.0;
        var logp = new double[maskIdx.Count];
        for (var c = 0; c < nCh; c++)
        {
            for (var j = 0; j < maskIdx.Count; j++)
                logp[j] = Math.Log10(Math.Max(psd[c][maskIdx[j]], 1e-20));
            var (slope, intercept) = LinFit(logf, logp);
            expSum += -slope;      // aperiodic exponent = negative log-log slope
            offSum += intercept;
        }
        return new AperiodicFit(expSum / nCh, offSum / nCh);
    }

    /// <summary>Ordinary-least-squares slope &amp; intercept (identical to aperiodic.py).</summary>
    private static (double Slope, double Intercept) LinFit(double[] xs, double[] ys)
    {
        var n = xs.Length;
        double mx = 0.0, my = 0.0;
        for (var i = 0; i < n; i++) { mx += xs[i]; my += ys[i]; }
        mx /= n; my /= n;
        double num = 0.0, den = 0.0;
        for (var i = 0; i < n; i++)
        {
            var dx = xs[i] - mx;
            num += dx * (ys[i] - my);
            den += dx * dx;
        }
        var slope = den != 0.0 ? num / den : 0.0;
        return (slope, my - slope * mx);
    }
}
